opsdroid/connector/twitch/__init__.py

```diff
@@ -0,0 +1,738 @@
+        """Send needed data to the websockets to be able to make a connection.
+        
+        If we try to connect to Twitch with an expired oauth token, we need to 
+        request a new token. The problem is that Twitch doesn't close the websocket
+        and will only notify the user that the login authentication failed after
+        we sent the `PASS`, `NICK` and `JOIN` command to the websocket.
+        
+        So we need to send the initial commands to Twitch, await for a status with
+        `await self.websockets.recv()` and there will be our notification that the 
+        authentication failed in the form of `:tmi.twitch.tv NOTICE * :Login authentication failed`
+        
+        This method was created to prevent us from having to copy the same commands
+        and send them to the websocket. If there is an authentication issue, then we
+        will have to send the same commands again - just with a new token.
+        
+        """
+        await self.websocket.send_str(f"PASS oauth:{self.token}")
+        await self.websocket.send_str(f"NICK {self.bot_name}")
+        await self.websocket.send_str(f"JOIN #{self.default_target}")
+        
+        await self.websocket.send_str("CAP REQ :twitch.tv/commands")
+        await self.websocket.send_str("CAP REQ :twitch.tv/tags")
+        await self.websocket.send_str("CAP REQ :twitch.tv/membership")
+
+    async def connect_websocket(self):
+        """Connect to the irc chat through websockets.
+        
+        Our connect method will attempt to make a connection to Twitch through the
+        websockets server. If the connection is made, any sort of failure received
+        from the websocket will be in the form of a `NOTICE`, unless Twitch closes
+        the websocket connection.
+        
+        In this method we attempt to connect to the websocket and use the previously
+        saved oauth token to join a twitch channel.
+        
+        Once we are logged in and on a Twitch channel, we will request access to special
+        features from Twitch.
+        
+        The `commands` request is used to allow us to send special commands to the Twitch
+        IRC server.
+        
+        The `tags` request is used to receive more information with each message received 
+        from twitch. Tags enable us to get metadata such as message ids.
+        
+        The `membership` request is used to get notifications when an user enters the 
+        chat server (it doesn't mean that the user is watching the streamer) and also when 
+        a user leaves the chat channel.
+        
+        """
+        _LOGGER.info(_("Connecting to Twitch IRC Server."))
+
+        async with aiohttp.ClientSession() as session:
+            async with session.ws_connect(f"{self.server}:{self.port}", heartbeat=600) as websocket:
+                self.websocket = websocket
+                await self.send_handshake()
+                await self.get_messages_loop()
```

Comment:
Apparently when I do `raise aiohttp.ServerDisconnectedError`  inside the `get_messages_loop` this will throw an "exception not caught", I would expect that the try/catch inside the `listen` method should handle that but doesn't seem like it.

So I might have to re-raise the same exception here, what do you think about it? 🤔

Reply:
Just another update, apparently trying to send `pong` to Twitch with `websocket.pong()` doesn't work because Twitch wants us to send it as a string. So changing the reply to `webscket.send_str("PONG")` seems to have fixed the issue of the connection dropping constantly around 6 minutes.

Now that makes me a bit afraid that the whole reconnect is still messed up and might not be able to catch the exception. But also makes me realise that probably the issue might have been happening because Twitch allowed us to reconnect to the service but then didn't get a proper reply and kind of put us on a timeout? So maybe the `aiohttp.ClientSession` request got an error message and since I'm not catching anything there it just stayed silet.

Anyway apologies for the rambling and I will need to test this further to see if we get that exception eventually. Will push the code to the digital ocean droplet and let it run for a while to test things out.